### PR TITLE
feat(flightsql): Avoid planning Flight queries twice

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/influxdata/influxdb_iox/pull/7280~

Closes https://github.com/influxdata/influxdb_iox/issues/7278

# Rationale

It is wasteful to create an entire plan just to compute the schema for each `GetFlightInfo` call, as the schema is always the same, as suggested by @mhilton . Hopefully this will reduce the latency of running queries via FlightSQL

# Changes

This PR avoids creating an entire plan just to compute the schema

It is covered by the tests added in https://github.com/influxdata/influxdb_iox/pull/7280

### CODE REVIEW

File: flightsql/src/planner.rs [hunk 1]
```diff
@@ -44,50 +49,20 @@ impl FlightSQLPlanner {
             FlightSQLCommand::CommandPreparedStatementQuery(handle) => {
                 get_schema_for_query(handle.query(), ctx).await
             }
-            FlightSQLCommand::CommandGetSqlInfo(CommandGetSqlInfo { info }) => {
-                let plan = plan_get_sql_info(ctx, info).await?;
-                // As an optimization, we could hard code the result
-                // schema instead of recomputing it each time.
-                get_schema_for_plan(plan)
+            FlightSQLCommand::CommandGetSqlInfo(CommandGetSqlInfo { .. }) => {
+                encode_schema(iox_sql_info_list().schema())
```
Comment (alamb) [inline flightsql/src/planner.rs:53]: There is not a whole lot to this PR -- it is basically hard coding the return schemas (and the tests ensure they are consistent with the actual data -- if any of the schemas are changed, the test fails)

File: flightsql/src/planner.rs [hunk 2]
```diff
@@ -269,6 +244,10 @@ async fn plan_get_catalogs(ctx: &IOxSessionContext) -> Result<LogicalPlan> {
     Ok(ctx.sql_to_logical_plan(query).await?)
 }
 
+/// The schema for GetCatalogs
+static GET_CATALOG_SCHEMA: Lazy<Schema> =
+    Lazy::new(|| Schema::new(vec![Field::new("catalog_name", DataType::Utf8, false)]));
```
Comment (appletreeisyellow) [inline flightsql/src/planner.rs:249]: I think this is more like a Rust question: does this code lazy load the schema value on the first access and remember it for the entire lifetime of the `GET_CATALOG_SCHEMA` variable?

**Review by appletreeisyellow — APPROVED** [review]
Overall this PR makes sense to me! Just one Rust question below

File: flightsql/src/planner.rs (hunk 2 above)
Comment (alamb) [reply flightsql/src/planner.rs:249]: > does this code lazy load the schema value on the first access and remember it for the entire lifetime of the GET_CATALOG_SCHEMA variable?

Yes, that is correct. Specifically, the lifetime of the `GET_CATALOG_SCHEMA` variable is "static" which in Rust parlance means "the entire life of the program" (aka it only gets cleaned up at program exit).

The docs describe how `Lazy` works in more detail:  https://docs.rs/once_cell/1.17.1/once_cell/sync/struct.Lazy.html

(not sure if you have seen https://docs.rs/ but I used it all the time when working with libraries)

**Pull request merged**